compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs

```diff
@@ -599,43 +599,41 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     #[instrument(level = "debug", skip(self))]
     pub(in super::super) fn resolve_generator_interiors(&self, def_id: DefId) {
         self.select_obligations_where_possible(false, |_| {});
+
         let generators = std::mem::take(&mut *self.deferred_generator_interiors.borrow_mut());
         debug!(?generators);
 
         // First, we remove all pending obligations that are stalled on unifying the generator
         // witnesses.  Those are stored to be processed by `check_generator_obligations`.
-        for &(expr_hir_id, _, interior) in generators.iter() {
-            let expr_def_id = self.tcx.hir().local_def_id(expr_hir_id);
-            debug!(?expr_def_id);
-            let obligations =
-                self.fulfillment_cx.borrow_mut().drain_obligations_stalled_on(interior);
-            let obligations = obligations.into_iter().map(|o| (o.predicate, o.cause)).collect();
-            debug!(?obligations);
-            self.typeck_results
-                .borrow_mut()
-                .generator_interior_predicates
-                .insert(expr_def_id, obligations);
-        }
 
         // Now that no obligation references the interior type variables, we unify them with the
         // `ty::GeneratorWitness` type, so downstream analyses know how to get the information.
-        for (expr_hir_id, body_id, interior) in generators {
-            let expr_def_id = self.tcx.hir().local_def_id(expr_hir_id).to_def_id();
+        for &(expr_hir_id, body_id, interior) in generators.iter() {
+            let expr_def_id = self.tcx.hir().local_def_id(expr_hir_id);
+            debug!(?expr_def_id);
+
             let substs = InternalSubsts::identity_for_item(
                 self.tcx,
-                self.tcx.typeck_root_def_id(expr_def_id),
+                self.tcx.typeck_root_def_id(expr_def_id.to_def_id()),
             );
 
-            let witness = self.tcx.mk_generator_witness(expr_def_id, substs);
+            let witness = self.tcx.mk_generator_witness(expr_def_id.to_def_id(), substs);
             let span = self.tcx.hir().body(body_id).value.span;
-            match self
+            let ok = self
                 .at(&self.misc(span), self.param_env)
                 .eq(self.tcx.mk_ty_var(interior), witness)
-            {
-                // FIXME is this correct?
-                Ok(ok) => self.register_infer_ok_obligations(ok),
-                Err(err) => panic!("Failed to unify generator interior type: {:?}", err),
-            }
+                .expect("Failed to unify generator interior type");
+            let mut obligations = ok.obligations;
+
+            obligations
+                .extend(self.fulfillment_cx.borrow_mut().drain_unstalled_obligations(&self.infcx));
```

Comment:
why would any unstalled obligations be in here considering `select_obligations_where_possible` was called at the start of this function?